### PR TITLE
8311922: [macOS] right-Option key fails to generate release event

### PR DESCRIPTION
**Pull request by DamonGuy**
Previously, a new key combination involving the option key was added to Aqua LAF for JTextAreas. In doing so, some code was removed that created this regression. The regression caused the right option key on Mac OS to incorrectly show another KeyPressed event instead of a KeyReleased event when pressing and releasing the key. Additionally, the location of the key was 'standard' instead of 'right'. Adding back the key mask and its following code resolves the issue and doesn't cause any other CI tests to fail.

The headful test included displays the key events as they're pressed. After the changes, the test correctly shows the right option key's KeyPressed and KeyReleased events and shows the location as 'right'.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311922](https://bugs.openjdk.org/browse/JDK-8311922): [macOS] right-Option key fails to generate release event (**Bug** - P3)

### Reviewers
 * [Harshitha Onkar](https://openjdk.org/census#honkar) (@honkar-jdk - Committer) ⚠️ Review applies to [d1b3ae5e](https://git.openjdk.org/jdk/pull/15432/files/d1b3ae5e5b95bceb99c2c114fb55de8feecf4300)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15432/head:pull/15432` \
`$ git checkout pull/15432`

Update a local copy of the PR: \
`$ git checkout pull/15432` \
`$ git pull https://git.openjdk.org/jdk.git pull/15432/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15432`

View PR using the GUI difftool: \
`$ git pr show -t 15432`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15432.diff">https://git.openjdk.org/jdk/pull/15432.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15432#issuecomment-1693743240)

### CODE REVIEW

**Comment by honkar-jdk** [comment]
@DamonGuy  Tested with the patch. Observed that the key code and extended modifier differ for left vs right option key. This is not the case for other left/right keys such as Shift, Command. I think some extra flags are being set in case of option key.

```
LEFT SHIFT KEY
-----------------
KEY PRESSED: 
 key code = 16 (⇧)
 extended modifiers = 64 (⇧)
 action key? NO
 key location: left
KEY RELEASED: 
 key code = 16 (⇧)
 extended modifiers = 0 (no extended modifiers)
 action key? NO
 key location: left

RIGHT SHIFT KEY
-----------------
KEY PRESSED: 
 key code = 16 (⇧)
 extended modifiers = 64 (⇧)
 action key? NO
 key location: right
KEY RELEASED: 
 key code = 16 (⇧)
 extended modifiers = 0 (no extended modifiers)
 action key? NO
 key location: right
------------------------------------------------------------------

LEFT OPTION KEY
------------------
KEY PRESSED: 
 key code = 18 (⌥)
 extended modifiers = 512 (⌥)
 action key? NO
 key location: left
KEY RELEASED: 
 key code = 18 (⌥)
 extended modifiers = 0 (no extended modifiers)
 action key? NO
 key location: left

RIGHT OPTION KEY
-------------------
KEY PRESSED: 
 key code = 65406 (⌥)
 extended modifiers = 8704 (⌥+⌥)
 action key? NO
 key location: right
KEY RELEASED: 
 key code = 65406 (⌥)
 extended modifiers = 0 (no extended modifiers)
 action key? NO
 key location: right

```

**Comment by prsadhuk** [comment]
I guess it was removed in as part of JDK-8267374 fix. Please see [PR#9230](https://github.com/openjdk/jdk/pull/9230) where it was concluded that we dont need ALT_GRAPH mapping for macos and RightOption key was mapped to ALT_GRAPH which makes Option key not work
https://github.com/openjdk/jdk/pull/9230#issuecomment-1165315655
https://github.com/openjdk/jdk/pull/9230#discussion_r906279285

Please run the regression test added/modified to see it does not regress those, mainy
test/jdk/javax/swing/JTextArea/TestAltUpDown.java

**Comment by DamonGuy** [comment]
> @DamonGuy Tested with the patch. Observed that the key code and extended modifier differ for left vs right option key. This is not the case for other left/right modifier keys such as Shift, Command. I think some extra flags (key masks) are being set in case of **Right Option Key**.
> 
> ```
> LEFT SHIFT KEY
> -----------------
> KEY PRESSED: 
>  key code = 16 (⇧)
>  extended modifiers = 64 (⇧)
>  action key? NO
>  key location: left
> KEY RELEASED: 
>  key code = 16 (⇧)
>  extended modifiers = 0 (no extended modifiers)
>  action key? NO
>  key location: left
> 
> RIGHT SHIFT KEY
> -----------------
> KEY PRESSED: 
>  key code = 16 (⇧)
>  extended modifiers = 64 (⇧)
>  action key? NO
>  key location: right
> KEY RELEASED: 
>  key code = 16 (⇧)
>  extended modifiers = 0 (no extended modifiers)
>  action key? NO
>  key location: right
> 
> ------------------------------------------------------------------
> 
> LEFT OPTION KEY
> ------------------
> KEY PRESSED: 
>  key code = 18 (⌥)
>  extended modifiers = 512 (⌥)
>  action key? NO
>  key location: left
> KEY RELEASED: 
>  key code = 18 (⌥)
>  extended modifiers = 0 (no extended modifiers)
>  action key? NO
>  key location: left
> 
> RIGHT OPTION KEY
> -------------------
> KEY PRESSED: 
>  key code = 65406 (⌥)
>  extended modifiers = 8704 (⌥+⌥)
>  action key? NO
>  key location: right
> KEY RELEASED: 
>  key code = 65406 (⌥)
>  extended modifiers = 0 (no extended modifiers)
>  action key? NO
>  key location: right
> ```

Thanks for catching this. I believe the edit I just pushed fixes this. My test now shows the correct key code and location on my end.

**Comment by DamonGuy** [comment]
> I guess it was removed in as part of JDK-8267374 fix. Please see [PR#9230](https://github.com/openjdk/jdk/pull/9230) where it was concluded that we dont need ALT_GRAPH mapping for macos and RightOption key was mapped to ALT_GRAPH which makes Option key not work [#9230 (comment)](https://github.com/openjdk/jdk/pull/9230#issuecomment-1165315655) [#9230 (comment)](https://github.com/openjdk/jdk/pull/9230#discussion_r906279285)
> 
> Please run the regression test added/modified to see it does not regress those, mainly test/jdk/javax/swing/JTextArea/TestAltUpDown.java

I see the comments. I removed the code that was removed again and instead edited the existing NSAlternateKeyMask. Fix now shows the correct key code as well. TestAltUpDown passes as well as my new test.

**Comment by alisenchung** [comment]
why was the keycode wrong with the previous fix? and why do we need to manually set leftAltKeyPressed = YES? should this be done in a different part of the code?

File: src/java.desktop/macosx/native/libawt_lwawt/awt/AWTEvent.m [hunk 1]
```diff
@@ -291,7 +291,7 @@
         //kCGSFlagsMaskAppleLeftAlternateKey,
         //kCGSFlagsMaskAppleRightAlternateKey,
         58,
-        0,
+        61,
```
Comment (honkar-jdk) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/AWTEvent.m:294]: Looks like the only change required to make right option key working is adding in the rightKeyCode (61) here. I don't think it is required to revert the `leftAltKeyPressed` changes.

File: src/java.desktop/macosx/native/libawt_lwawt/awt/AWTEvent.m [hunk 2]
```diff
@@ -309,6 +309,7 @@
     {0, 0, 0, 0, 0, 0}
 };
 
+static BOOL leftAltKeyPressed;
```
Comment (honkar-jdk) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/AWTEvent.m:312]: I think all the occurrences of `leftAltKeyPressed` in AWTEvent.m can be removed.

**Review by honkar-jdk — CHANGES_REQUESTED** [review]

File: src/java.desktop/macosx/native/libawt_lwawt/awt/AWTEvent.m (hunk 1 above)
Comment (DamonGuy) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/AWTEvent.m:294]: This seems correct. I updated it to only include the key code. Re-ran my test and TestAltUpDown, and both passed still.

**Comment by DamonGuy** [comment]
> why was the keycode wrong with the previous fix? and why do we need to manually set leftAltKeyPressed = YES? should this be done in a different part of the code?

I originally added back the code that was removed. There seemed to be 2 instances of NSAlternateKeyMask afterwards, so the key code value was incorrect. This must've been the case before the regression as well I'd assume. leftAltKeyPressed has been removed as well since it doesn't seem necessary to fix this particular issue.

File: test/jdk/java/awt/event/KeyEvent/OptionKeyEventTest.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+import javax.swing.JButton;
+import javax.swing.JFrame;
+import javax.swing.JScrollPane;
+import javax.swing.JTextArea;
+import javax.swing.JTextField;
+import javax.swing.SwingUtilities;
+
+public class OptionKeyEventTest extends JFrame
+        implements KeyListener, ActionListener
+{
```
Comment (honkar-jdk) [inline test/jdk/java/awt/event/KeyEvent/OptionKeyEventTest.java:52]: Formatting:  '{' should be on the previous line

Comment (DamonGuy) [reply test/jdk/java/awt/event/KeyEvent/OptionKeyEventTest.java]: Fixed

**Review by honkar-jdk — APPROVED** [review]
LGTM. Manual test looks good. 

I guess this test cannot be automated using Robot since the `VK_ALT` defaults to the left option key.

**Comment by DamonGuy** [comment]
> LGTM. Manual test looks good.
> 
> I guess this test cannot be automated using Robot since the `VK_ALT` defaults to the left option keypress.

Right. I looked into automating this and making the test headless, but I didn't see a way around this. Thanks

**Review by prrace — APPROVED** [review]

**Pull request closed without merging**